### PR TITLE
Propagate anchors to marks, not just ligature of marks

### PR DESCRIPTION
**Pull request by moyogo**
The PropagateAnchorFilter only propagate anchors to ligature of marks (for example circumflexcomb_macroncomb).
With this change, the anchors are also propagate to other composites of marks (for example macroncomb.alt).

Glyphs.app (2 and 3) propagate anchors to composite of marks, not just ligatures of marks.

### CODE REVIEW

File: Lib/ufo2ft/filters/propagateAnchors.py [hunk 1]
```diff
@@ -76,17 +76,18 @@ def _propagate_glyph_anchors(glyphSet, composite, processed):
...
+            # "circumflexcomb_tildecomb"). Promote the mark that is positioned
+            # closest to the origin to a base.
+            try:
+                component = _component_closest_to_origin(mark_components, glyphSet)
+            except Exception as e:
+                raise Exception(
+                    "Error while determining which component of composite "
+                    "'{}' is the lowest: {}".format(composite.name, str(e))
+                )
         mark_components.remove(component)
```
Comment (anthrotype) [inline Lib/ufo2ft/filters/propagateAnchors.py:91]: where is the local variable `component` being defined/assigned? I only see an explicit assignment inside the `if _is_ligature_mark(composite)` block. What if the if statement is false?

Comment (madig) [reply Lib/ufo2ft/filters/propagateAnchors.py:91]: Probably needs a `else: component = composite`?

Comment (moyogo) [reply Lib/ufo2ft/filters/propagateAnchors.py]: Ouch I should have seen that. Fixed in new commits.

File: Lib/ufo2ft/filters/propagateAnchors.py [hunk 2]
```diff
@@ -76,26 +76,35 @@ def _propagate_glyph_anchors(glyphSet, composite, processed):
                 base_components.append(component)
                 anchor_names |= {a.name for a in glyph.anchors}
```
Comment (madig) [inline Lib/ufo2ft/filters/propagateAnchors.py:77]:
```suggestion
                anchor_names.update(a.name for a in glyph.anchors)
```

File: Lib/ufo2ft/filters/propagateAnchors.py [hunk 3]
```diff
@@ -76,26 +76,35 @@ def _propagate_glyph_anchors(glyphSet, composite, processed):
...
+            base_components.append(component)
+            glyph = glyphSet[component.baseGlyph]
+            anchor_names |= {a.name for a in glyph.anchors}
 
     for anchor_name in anchor_names:
         # don't add if composite glyph already contains this anchor OR any
         # associated ligature anchors (e.g. "top_1, top_2" for "top")
-        if not any(a.name.startswith(anchor_name) for a in composite.anchors):
+        if not any(
+            a.name == anchor_name or a.name.split("_")[0] == (anchor_name)
```
Comment (madig) [inline Lib/ufo2ft/filters/propagateAnchors.py:105]: What does this catch that the previous code didn't? 🤔

Comment (moyogo) [reply Lib/ufo2ft/filters/propagateAnchors.py:105]: "top" would not be propagated if "topright" is present because `"topright".startswith("top")`.

**Comment by madig** [comment]
Note to self: see if it possible to get rid of recursion by using something like graphlib: https://github.com/madig/HTLetterspacer/blob/ed78cd1a8096f7bfaeb3ec73d83c463aca22fe69/src/htletterspacer/__main__.py#L69-L84

File: Lib/ufo2ft/filters/propagateAnchors.py [hunk 4]
```diff
@@ -76,26 +76,35 @@ def _propagate_glyph_anchors(glyphSet, composite, processed):
...
+                    "'{}' is the lowest: {}".format(composite.name, str(e))
+                )
+            components = [component]
+        else:
+            components = mark_components
+
+        for component in components:
+            mark_components.remove(component)
+            base_components.append(component)
+            glyph = glyphSet[component.baseGlyph]
```
[the remark this answers is not in the document]
Comment (madig) [reply Lib/ufo2ft/filters/propagateAnchors.py:98]: Nevermind, we recurse into all components before looking at their anchors.

File: Lib/ufo2ft/filters/propagateAnchors.py [hunk 5]
```diff
@@ -74,28 +74,37 @@ def _propagate_glyph_anchors(glyphSet, composite, processed):
...
+            try:
+                component = _component_closest_to_origin(mark_components, glyphSet)
+            except Exception as e:
+                raise Exception(
+                    "Error while determining which component of composite "
+                    "'{}' is the lowest: {}".format(composite.name, str(e))
+                )
+            components = [component]
+        else:
+            components = mark_components
```
Comment (madig) [inline Lib/ufo2ft/filters/propagateAnchors.py:93]: What is the logic behind promoting _all_ mark components to bases here? This works for single component glyphs like the `macroncomb.alt` in your test but what if it's not so simple?

**Comment by moyogo** [comment]
graphlib looks interesting.

I think we may have to refactor more than just this to do, especially to do some of what is described in https://github.com/googlefonts/glyphsLib/issues/368#issuecomment-491939617.

**Comment by moyogo** [comment]
I have another branch that will supersede this one.

**Pull request closed without merging**